### ISSUE TITLE
Extension Types

### ISSUE THREAD

**Issue by findepi** [apache/datafusion#12644]
### Is your feature request related to a problem or challenge?

Currently DataFusion provides a lot of built-in types which are useful when building applications / query engines on top of DataFusion. However, even plethora of types is not enough. DataFusion doesn't have types existing in other systems, limiting DataFusion applicability as "LLVM for query engines"

For example, these types commonly found in other systems do not exist today
- char(n)
- varchar(n)
- timestamp with time zone (a pair of "point in time" + "time zone" information; found in Oracle, Trino, Snowflake, etc.)
  - DataFusion currently uses Arrow `DataType` and the closest Arrow has is "timestamp(zone)" where each value is in _same_ zone
- timestamp with local time zone (point in time without zone information; found in Spark, Hive, PostgreSQL)
  - DataFusion currently uses Arrow `DataType` and the closest Arrow has is "timestamp(zone)" with eg UTC zone. however cast to varchar for "timestamp(UTC)" and for "timestamp with local time zone" should behave differently
- time with time zone 
- JSON
  - DataFusion currently uses Arrow `DataType` and the closest Arrow has Utf8 potentially with some metadata information. Utf8 might be a perfect carrier type for JSON data, but "cast(json AS T)" and "cast(utf8 AS T)" are usually pretty different operations
- VARIANT (https://github.com/apache/datafusion/issues/10987)
- geospatial Geometry types (https://github.com/apache/datafusion/issues/7859)
- HLL (hyperloglog), digests (t-digest, q-digest, other statistical digests)
- extensions for applications building on top of DF; including user defined types (UDT) (https://github.com/apache/datafusion/issues/7923)
  - ability to provide _user_-defined types is even broader than ability to provide extension types ("_rust_-defined types")

### Describe the solution you'd like

1. Introduction of DataFusion own type system
   - this is generally covered by https://github.com/apache/datafusion/issues/11513 & https://github.com/apache/datafusion/issues/12622
2. Introduction of extensions in DataFusion type system allowing applications building on DataFusion to provide more types 
   - the extension types -- not unlike DataFusion built-in types -- need to use Arrow types as "carrier type" for transporting
   - the Arrow type metadata weaved into schema _fields_ can be used to indicate use of extension types to the client, when data is returned to the user in Arrow form
   - for example, a "timestamp with time zone" type _could_ be represented as Struct with two fields: point_in_time, time_zone
4. Ability to dynamically find operations on types during function resolution or runtime
   - for example a `CAST(array<T> AS varchar)` needs to know how to do `cast(T AS varchar)`. It cannot delegate this logic fully to Arrow, because Arrow won't have a notion of extension types.
     - eg if "timestamp with time zone" uses a Struct as a carrier type, it still needs to define its own `cast(... AS varchar)`. It cannot use the default `cast(struct AS varchar)`.

#### Everything is built-in

DataFusion could provide all types needed by applications building on top of DataFusion as built-in DataFusion types.
This would be easiest to implement, but could lead to scope-creep for the project. This could also lead to conflicts where types look the same but the desired behavior differs between applications building on top of DataFusion. For example Oracle's and Trino's "timestamp with time zone" can represent political zones while Snowflake's allows only fixed offsets.

#### No-op

Not providing extension types. This would limit DataFusion applicability.
DataFusion cannot be considered "LLVM for query engines" if it cannot serve as an engine, or potential engine, for existing popular query engines. 

### Additional context

The need to create extension types was raised in the [Proposal] Decouple logical from physical types

- https://github.com/apache/datafusion/issues/11513 & https://github.com/apache/datafusion/issues/12622

However introduction of DataFusion own types does not require introduction of extension types.
Extension types are complex enough (especially given their impact on functions) that they deserve their own roadmap issue.

The impact of extension types on functions, functions runtime and resolution is very clear, so this relates to Simple Functions initiative:

- https://github.com/apache/datafusion/issues/12635

**Comment by findepi** [comment]
cc @alamb, @andygrove, @jayzhan211, @ozankabak, @notfilippo, @comphead, @kylebarron, @yukkit, @sunchao, @Folyd, @wjones127, @Xuanwo, @sadboy, @milevin

**Comment by kylebarron** [comment]
I'm not very knowledgeable about DataFusion internals or database theory, so it's hard for me to provide feedback on the proposal, but I'm very excited about the prospect of extension types to enable spatial types (https://github.com/apache/datafusion/issues/7859). I've been collaborating on the [GeoArrow](https://geoarrow.org/) spec, which defines Arrow extension types for spatial data. It's important to have additional logical types because the same physical layout can be interpreted in multiple logical ways (e.g. an array of `LineString` and `MultiPoint`), and to store [coordinate reference system](https://en.wikipedia.org/wiki/Spatial_reference_system) information (what physical locations on earth these numbers represent) as part of the type.

**Comment by findepi** [comment]
FYI i touched upon the topic of types on DataFusion meetup in Belgrade yesterday.
The slides are here if anyone is interested: https://docs.google.com/presentation/d/1VW_JCGbN22lrGUOMRvUXGpAmlJopbG02hn_SDYJouiY . It was an attempt to summarize why we need both: simpler types (https://github.com/apache/datafusion/issues/11513), more types (https://github.com/apache/datafusion/issues/12644), and simple function "SDK" (https://github.com/apache/datafusion/issues/12635).
The document has comments disabled to avoid diverging the discussion from the github issue.

**Comment by alamb** [comment]
See a previous proposal from @yukkit : https://github.com/apache/datafusion/issues/7923

**Comment by paleolimbot** [comment]
I'm interested in this as well; however, I'm new to DataFusion development and I am not sure I have a handle on what the barriers are here (e.g., Is support for this blocked by lack of support in arrow-rs? Is there opposition to this on principle or is it just a matter of development time?).

A few more (albeit geospatial) reasons to add support for this are (not-quite-merged!) support for spatial data in Iceberg and Parquet. As currently written, those proposals include type-level metadata (notably: the CRS), which would mean that reading Parquet files with those types would loose information. There are also other [Arrow canonical extension types](https://arrow.apache.org/docs/format/CanonicalExtensions.html) that have type-level metadata (which to my current reading, means that canonical extension types are not usable in DataFusion?).

**Comment by kylebarron** [comment]
> Is support for this blocked by lack of support in arrow-rs? 

arrow-rs intentionally doesn't "directly support" extension types through `DataType`. The alternative (now archived) [arrow2 implementation](https://github.com/jorgecarleitao/arrow2) had [an `Extension` variant](https://docs.rs/arrow2/latest/arrow2/datatypes/enum.DataType.html#variant.Extension), but this has the downside of requiring _all_ code that needs to match the `DataType` enum to be aware of logical extension types. See https://github.com/apache/arrow-rs/issues/4472 for more discussion.

Instead, arrow-rs supports extension types indirectly via the metadata hashmap on the `Field` accompanying some `Array`. So e.g. in geoarrow-rs, we have custom types like `PointArray` that internally manage both the array data and the field metadata, but when exporting to `arrow` objects, you'd need to carry the `Array` and the `Field` together.

I haven't closely followed the logical type work in progress in Datafusion but I assume it would be associating some Field with the physical type, and extension types could inject metadata there.

**Comment by tobixdev** [comment]
I second @paleolimbot here, as I am also interested in this topic.
First of all, thank you for all your efforts in this area!

To provide some context, we are currently creating a prototype that handles data that is encoded as a union with 10-ish different variants. Logically, this union represents the encoding of a single type. While we make ends meet by creating logical plans manually with specialized UDFs (e.g., for equality),  we are in a very early phase. Only working with physical types will add lots of complexity to our code that can be eliminated with [logical types](#12622) and its benefits (e.g., #12635).

So what our (dream) scenario is that we could define this extension type (and its encoding) together with implementations for equality, orderings, etc, and DataFusion would automatically make use of these when joining or sorting (AFAIK, this should be supported according to #7923).

I'd like to support you in these efforts. However, while I read a few discussions and proposals, I am still a bit lost on where I can/should help out as this is (from what I see) a huge ongoing project across multiple issues and epics. Do you have pointers on where I could start helping out?  For example, I found #13301, but I am unsure if the efforts on simple functions #12635 make these changes somewhat premature.

Thank you for helping me to navigate this project!

**Comment by jayzhan-synnada** [comment]
The current status is that we have several changes in branch `logical-types` for #12622. Where the `Scalar` is introduced and the next step is to complete the tasks left in #12622. 

Remove ScalarValue::LargeUtf8/Utf8View/LargeBinary/BinaryView.

**Comment by jayzhan211** [comment]
The current status is that we have several changes in branch `logical-types` for #12622. Where the `Scalar` is introduced and the next step is to complete the tasks left in #12622. 

Remove ScalarValue::LargeUtf8/Utf8View/LargeBinary/BinaryView.

The change in `logical-types` is huge

**Comment by alamb** [comment]
@mbrobbel  has a nice proposal to add extension types in arrow-rs (which would potentially help ExtensionTypes in DataFusion). Would apprecaite any feedback: 
- https://github.com/apache/arrow-rs/pull/5822

**Comment by alamb** [comment]
I started thinking about how the ExtensionType trait would be used in DataFusion. I think we would need to improve the APIs a bit to be in terms of `Field` rather than `DataType`. See this ticket for more info
- https://github.com/apache/datafusion/issues/14247

**Comment by alamb** [comment]
Coming soon (arrow 54.2 in a month), support for Extension Types:
- https://github.com/apache/arrow-rs/issues/4472
- https://github.com/apache/arrow-rs/pull/5822
Courtesy of @mbrobbel  🎩

**Comment by paleolimbot** [comment]
Just a note to thank you for coordinating all of us on this effort! I am new to the code base and I don't have a great handle on exactly where to start, but I'm particularly interested in getting the extension type (or Field) propagated through to user-defined functions.

**Comment by tobixdev** [comment]
Here is a PR discussing a possible implementation for user-defined sorting: [15106](https://github.com/apache/datafusion/pull/15106) (more infos in the PR)

Could be interesting for folks interested in the extension types / user-defined type topic.

**Comment by tv42** [comment]
Related work done elsewhere: https://docs.vortex.dev/concepts/dtypes

**Comment by alamb** [comment]
Another thing that is probably worth mentioning is that I know of several efforts to use the Arrow "Extension Type" mechanism in DataFusion for user defined logical types
- Geometry (@paleolimbot ) 
- Something custom (@timsaucer )
- Variant (myself, @PinkCrow007, @mkarbo, @scovitch, and others)


Specifically; https://docs.rs/arrow-schema/55.1.0/arrow_schema/extension/trait.ExtensionType.html

**Comment by kylebarron** [comment]
I'm also working on geospatial support with @paleolimbot. For example there are GeoArrow types that implement `ExtensionType` in the [`geoarrow-schema` crate](https://docs.rs/geoarrow-schema/latest/geoarrow_schema/)

**Comment by alamb** [comment]
@paleolimbot and @timsaucer  -- @Omega359 and I have added a callout for this feature (using extension types) in the DataFusion 49.0.0 blog post here:
- https://github.com/apache/datafusion-site/pull/91

Do we know what the next steps are needed to help drive along  "user defined types" support?

**Comment by paleolimbot** [comment]
I think it's matter of what "support" entails...with metadata support it's certainly possible for metadata-based types like Arrow extension types to exist and for people to write functions that consume and produce them. There are also opportunities to support extension types in built-in DataFusion operations like signatures, casting, UNION ALL, pretty printing in the CLI, CSV output, SQL parsing/unparsing, and maybe a few I'm forgetting.

It's possible for an engine based on DataFusion to implement all of those today, but also DataFusion could take responsibility for some or all of those. Other engines (I'm thinking of DuckDB and Arrow C++) do this by creating an "extension" or "user defined" type object that defines behaviour that needs to be used by internals. In DuckDB this is just registering casts...in Pyarrow this is pretty much just defining type equality and type display. The [vctrs package in R](https://vctrs.r-lib.org/articles/s3-vector.html) allows defining a few more things like defining how arithmetic is done and how to concatenate an array with some other arbitrary type.

I'm happy to help with any of that but also I feel like I'm a bad guesser with respect to how the DataFusion/arrow-rs community would like to handle extensibility 🙂

**Comment by adriangb** [comment]
> it's certainly possible for metadata-based types like Arrow extension types to exist and for people to write functions that consume and produce them

Is it fully? For example could one make a UUID type stored as a fixed length binary that works with `col = 'abc'` as well as `col = 'a-b-c'`?

**Comment by paleolimbot** [comment]
> For example could one make a UUID type stored as a fixed length binary that works with col = 'abc' as well as col = 'a-b-c'?

If that example is SQL then the SQL parser would need some way to access a function that can parse a string of a user-registered type, so not today. If you're generating expressions using something else (e.g., Python), you can parse literals however you want or create functions that do the casting you want. DuckDB has a registry of casts that extensions can insert into and uses to/from string for this sort of thing.

**Comment by alamb** [comment]
My suggestion (somewhat self serving) is to do some "Demo Driven Development" here

Specifically, I suggest we try and code up an example (in `datafusion-exampes`) for a simple usecase (like UUID):
1. Define a Arrow extension type for UUID backed by a `FixedSizeBinaryArray`
3. Demonstrate how to read/write  "UUIDs" (ideally show reading/writing to parquet or some other file format)
2. Implement one or more functions that reads/writes from (only) UUIDs
4. Demonstrate running queries with UUIDs

If we hit issues / rough edges with that example, then we can file tickets / fix them. Once we have the example working then we could be confident saying "it is supported"

We could also use this example as part of a blog post on the topic:
- https://github.com/apache/datafusion-site/pull/73 (that @timsaucer and @paleolimbot are working on)

**Comment by adriangb** [comment]
> My suggestion (somewhat self serving) is to do some "Demo Driven Development" here

Agreed, I like this approach and your plan!

**Comment by findepi** [comment]
metadata-based types like Arrow extension types is definitely a viable and incrementally executable path.
However, the end result is remotely far from optimal: there is official type system represented by `DataType` and there is _actual_ type system represented by `DataType + properties bag`. Eventually, every use of `DataType` will need to be revisited and updated.

> For example could one make a UUID type stored as a fixed length binary that works with col = 'abc' as well as col = 'a-b-c'?

This is good example, and it's easy to create many others like this.
JSON backed by either Utf8 or Binary, what's the equality function? byte-for-byte, or structural?
VARIANT backed by Binary, containing serialized data. The equality _must_ deserialize.
UUID backed by  ...
In none of these cases the backing type (Utf8 or Binary) has any meaning and cannot be used for anything. Yet, it's a valid `DataType` so it _can be (erroneously) used for something_, be it an optimizer rule, a function call, a coercion logic, etc.

Is this the end state we want to achieve?

**Comment by paleolimbot** [comment]
I agree that DataFusion would benefit from embracing its own definition of a (dynamic, session-registered) data type that allows a type author to specify behaviour in such a way that DataFusion internals (e.g., optimizer rules, casting) can benefit things that are not just built-in arrow types (and/or prevent DataFusion internals from introducing correctness issues that may arise from ignoring or dropping metadata). I'm also sensing resistance to take on responsibility for that, which is OK (engines that use DataFusion can implement all of this on top of metadata, although it is quite a bit of work that perhaps doesn't have to be duplicated if it's in DataFusion).

**Comment by alamb** [comment]
> Is this the end state we want to achieve?

In my mind it is better than what is currently possible. I agree we can come up with other better designs too

> I agree that DataFusion would benefit from embracing its own definition of a (dynamic, session-registered) data type that allows a type author to specify behaviour in such a way that DataFusion internals (e.g., optimizer rules, casting) can benefit things that are not just built-in arrow types 

I mean I agree with this too  -- in my mind the question is how actually to achieve this goal

**Comment by findepi** [comment]
I don't know how to achieve `DataType + metadata`-based design without making things complicated and error-prone.
I think the bigger question is whether we as community agree regarding the end state we want to achieve. If we agree the end state is lean and pleasurable-to-work-with type system, I am sure we collectively figure it out how to get there.

**Comment by alamb** [comment]
> If we agree the end state is lean and pleasurable-to-work-with type system, I am sure we collectively figure it out how to get there.

I think it would he hard to find someone that doesn't want that  end state 😆 . I think the differences might be in opinions on what that actually looks like and how to get there

**Comment by paleolimbot** [comment]
For what it's worth, it looks like polars is implementing this as an `Extension` member on their `DataType` enum:

https://github.com/geopolars/geopolars/issues/245#issuecomment-3422086194

Having something `Arc<dyn ExtensionTypeyThing>` is nice because it limits the points at which a "registry" is necessary. The registry (ideally in the session) produces objects that know how to print, cast, coerce, and compare equality so that we don't have to pipe a reference to the registry into every single `==` or cast involving data types. In Arrow C++ the only points at which the static registry are used are when importing serialized IPC or importing a type via the C data interface.

**Comment by tschwarzinger** [comment]
> Having something Arc<dyn ExtensionTypeyThing> is nice because it limits the points at which a "registry" is necessary.

That's a valid point. If we further believe that all the metadata handling should be attached to `Field`, maybe the field should not directly have a physical `DataType`.  More like something akin to:

```rust
enum FieldDataType {
    Native(DataType),
    Extension(Arc<dyn ExtensionTypeyThing>)
}
```

This could i) allow arrow kernels to keep on focusing on `DataType` to interpret the underlying physical layout (which I think was the primary reason for rejecting the `DataType::Extension(...)` solution and ii) prevent accidentally using the wrong implementation of, for example, the `+` operator for an extension type as the type information stems from the `Field`. Again, a huge breaking change but I think this cannot be prevented. We could offer a utility functions that allows ignoring any logical type to ease the migration.

Any thoughts on that?  

This would be a similar approach to `TypeSignature`(https://github.com/apache/datafusion/blob/main/datafusion/common/src/types/logical.rs) .

**Comment by alamb** [comment]
I started a new ticket to focus on the  discussion about a registry of types / what exactly this API would look like in DataFusion. 
- https://github.com/apache/datafusion/issues/18223

I like the idea of this (maybe it is a new `DFType` 🤔 )

```rust
enum DFType {
    Native(DataType),
    Extension(Arc<dyn ExtensionTypeyThing>)
}
```

**Comment by alamb** [comment]
@paleolimbot and @tobixdev @timsaucer   -- I feel like there is significant motion / momentum for better Field / metadata support in service of user defined types, but I couldn't find any high level discussion about it other than this ticket

Do you know of anything more specific? Basically I am looking for a "epic" style ticked with a high level description of what we are trying to do (aka ensure `Field` is present everywhere instead of just DataType) and then a list of tickets for different areas that are needed 

If no such thing exists, would it be ok if I created one?

**Comment by paleolimbot** [comment]
> I feel like there is significant motion / momentum for better Field / metadata support in service of user defined types

Glad to hear it!

> Do you know of anything more specific?

I don't! There is some good discussion on https://github.com/apache/datafusion/issues/18223 with respect to some first steps.

> what we are trying to do (aka ensure Field is present everywhere instead of just DataType) 

I am not sure that is the consensus on this ticket or #18223 so far...I think we all want something better than that but aren't sure how to do it without changing arrow-rs or a lot of disruption of the DataFusion code base. Replacing the DataType with a Field in the logical plan and SQL parser is sort of the bare minimum for somebody else to invent their own type system on top of DataFusion with some effort.

> If no such thing exists, would it be ok if I created one?

That would be great! I am definitely willing to do work, it's just hard to do so without consensus and I am not sure what that is. My personal first step might be to replace any `FieldRef` that is meant to communicate a data type with `ExprType` or `ExprField` or `DFField` (that is a thin wrapper around a `FieldRef`) so that we can get metadata everywhere in the short term with the flexibility to update the internal representation (e.g., with a `dyn` type of some kind).

**Comment by alamb** [comment]
> That would be great! I am definitely willing to do work, it's just hard to do so without consensus and I am not sure what that is.

Yeah I agree that is key missing part. Let me see if I can write up something that maybe will help us get there. Will try tomorrow

**Comment by tschwarzinger** [comment]
I basically agree with @paleolimbot  here! I've also now opened a "discussion PR" for the extension type registry #18552 . Maybe this provides additional context.
